docs/src/cli/deploy-a-program.md

```diff
@@ -297,6 +308,74 @@ the program:
 solana program deploy --program-id <PROGRAM_ADDRESS> --buffer <BUFFER_ADDRESS>
 ```
 
-Note, the buffer's authority must match the program's upgrade authority.
+Note, the buffer's authority must match the program's upgrade authority. Also, upon successful deploy
+buffer accounts contents are copied into program accounts and are erased from blockchain.
 
 Buffers also support `show` and `dump` just like programs do.
+
+### Upgrading program using offline signer as authority
+
+Storing private key(s) on machine without internet access (offline signer) is much safer compared to
+storing them on machine with internet access (online signer), this section describes how program developer
+can use offline machine (offline signer) to sign-off on his program `upgrade` to get a **higher degree of
+security** (compared to typical upgrade flow described in [this section](deploy-a-program.md#redeploy-a-program)
+which assumes the usage of machine connected to internet - aka online signer). Note, currently only `upgrade`
+operation can be performed in "offline" mode, initial program `deploy` **must** be performed using online machine
+and only subsequent program upgrades can leverage offline signing. For a first time deploy it is **recommended** 
+to [change program's upgrade authority (using `--skip-new-upgrade-authority-signer-check` option)](deploy-a-program.md#set-a-programs-upgrade-authority)
```

Comment:
> We should also support offline signing for set-upgrade-authority too.

Yep, do you prefer a separate follow-up PR for this ^ (seems cleaner to merge it separately) ?

> From a safety perspective, it's not great to use --skip-new-upgrade-authority-signer-check since there's a possibility of getting the address wrong.

Right, I've mentioned this in https://github.com/norwnd/solana/blob/program-sign-only/docs/src/cli/deploy-a-program.md#set-a-programs-upgrade-authority, can't think of a simple way to address this, but one option would be to:
- instead of using `--skip-new-upgrade-authority-signer-check` flag require a signature from `new-upgrade-authority` (not just his pubkey) - which essentially means User "proves" to Cli he is indeed in control of new upgrade authority - which should address this issue, at least to some extent

Any thoughts on this ? Perhaps we'd want to create an issue for this so we don't forget about it.

Reply:
> Yep, do you prefer a separate follow-up PR for this ^ (seems cleaner to merge it separately) ?

Yeah that'll be much better, thanks, I created #34312 to track it